### ISSUE TITLE
Finding optimal parameters for a simple SIR model

### ISSUE THREAD

**Issue by abrahami** [AgentTorch/AgentTorch#30]
Hi there,
I am trying to use the package for a relatively simple scenario:
I have a network where nodes are users, and edges represent _possible_ interactions between users (in each episode). All edges in the network have the same infection parameter probability, which is unknown (i.e., **beta**).

I know the initial infected users (seed users) and the number of infected nodes by the end of the process. I wish to find the optimal **beta** parameter given this information. I assume the infection model follows the SIR model.

I was able to run the COVID example with the default setup suggested in the repo. However, I am struggling with two things:
1. How can I adjust/add this beta parameter to the model? Does the existing COVID model assume this parameter exists in the network and find its optimal value?
2. The example [here](https://github.com/AgentTorch/AgentTorch/blob/master/agent_torch/models/covid/main.py) only shows how to run the simulation, but further analysis of the model is not presented. How can I access the learned parameters for example?

Thank you!

**Comment by gamemaker1** [comment]
Hi @abrahami,

> 1. How can I adjust/add this beta parameter to the model? Does the existing COVID model assume this parameter exists in the network and find its optimal value?

I don't think the current model uses the beta parameter, although it does appear in the configuration [here](https://github.com/AgentTorch/AgentTorch/blob/40dde09f7f530aec27c493ab797e45a829ae25d2/agent_torch/models/covid/yamls/config.yaml#L28-L31). You could add it as a learnable argument in the SubstepTransition for the `new_transmission` substep, like [this](https://github.com/AgentTorch/AgentTorch/blob/40dde09f7f530aec27c493ab797e45a829ae25d2/agent_torch/models/covid/yamls/config.yaml#L410-L418), and then modify the substep code to use the parameter in the simulation.

> 2. The example [here](https://github.com/AgentTorch/AgentTorch/blob/master/agent_torch/models/covid/main.py?rgh-link-date=2024-07-01T16%3A16%3A16Z) only shows how to run the simulation, but further analysis of the model is not presented. How can I access the learned parameters for example?

The `Runner` is a `torch.nn.Module`, which means all the learnable parameters can be accessed by running `runner.parameters()`. You can then optimize those parameters for the next episode/epoch:

```py
num_episodes = self.config["simulation_metadata"]["num_episodes"]
num_steps_per_episode = self.config["simulation_metadata"]["num_steps_per_episode"]
optimizer = SomeOptimizer(runner.parameters(), ...)

for episode in range(num_episodes):
    optimizer.zero_grad()
    runner.reset()

    runner.step(num_steps_per_episode)
```

Hope this helps.

**Comment by abrahami** [comment]
Thank you @gamemaker1 ! It is a much better starting point now.
Another follow-up question, please:
Given that my data contains only the initial infected users (seed users) and the final ones (without any information about the people infected in between these two stages) -- would AgentTorch be a good fit to find the optimal parameters in such settings?

**Comment by gamemaker1** [comment]
It might be a bit difficult, you'll have to ensure that after all the substeps are run you get the exact same number of people infected finally.

For the optimizing parameters bit, yes, definitely, AgentTorch is a good fit.

**Comment by abrahami** [comment]
Well, the concept is to optimize the parameters of the model to ensure that the ending number of infected users is as close as posibble to the right number. Another words, the loss take into acount this objective.

Up to now we did so in a grid search fashion. We found the 'right' value by searching the one which yields the closest number of infected users by the end of a simulation run.
Should we try and use the package to do the same but in a faster way (and a different approach)?

**Comment by gamemaker1** [comment]
> Well, the concept is to optimize the parameters of the model to ensure that the ending number of infected users is as close as possible to the right number. In other words, the loss function should take into acount this objective.

If it can be incorporated into the loss function, I think AgentTorch will work pretty well to find the optimal value of parameters.
